### PR TITLE
Add CHANGELOG validation check

### PR DESCRIPTION
**Pull request by xRuiAlves**
Given that we follow some strict-ish structure for the Changelog, thought we might as well have some validation around it!

Let me know if you think this is helpful and, if so, if there are other checks that would be cool to include as well here.

**Side Note:** I also noticed that some files had a bunch of trailing whitespaces at EOLs, ended up removing them because my editor does it for me 😅

### Does this change relate to existing issues or pull requests?

No.

### Does this change require an update to the documentation?

No.

### How has this been tested?

Tested local using `act -j validate-changelog`, and in CI by... using CI 😛 

#### Using 'Changing' as a header instead of 'Changed'

```
Found invalid changelog sections! The invalid sections are:
### Changing
```

#### Using an invalid date format (`2023-9-07` instead of `2023-09-07`)

```
Found invalid changelog version headers! The invalid version headers are:
## [0.19.0] - 2023-9-07
```

#### Using an invalid version format (`0.19.0.1`)

```
Found invalid changelog version headers! The invalid version headers are:
## [0.19.0.1] - 2023-10-07
```

### CODE REVIEW

**Review by jcazevedo — DISMISSED** [review]
I'm not super fond of this, but I'm open to get it merged and see where that leads us. 

My main concern is with adding a non-negligible piece of code to validate something that is very low risk. If something is off in the Changelog it can quickly be fixed in a non-breaking way. We're also validating against things that are easy to spot in code reviews and being strict on a file that can be seen as a draft until an actual release.

**Comment by xRuiAlves** [comment]
> I'm not super fond of this, but I'm open to get it merged and see where that leads us.
> 
> My main concern is with adding a non-negligible piece of code to validate something that is very low risk. If something is off in the Changelog it can quickly be fixed in a non-breaking way. We're also validating against things that are easy to spot in code reviews and being strict on a file that can be seen as a draft until an actual release.

Yeah, you do have a good point. 
My motivation was mainly due to having to rely on said code reviews in the past (in other projects, not in this) proved to be a bit faulty at some times.
Given that this is an open source project, I think it might save some re-iterating time in having automatic validation in the overall structure.

@jcazevedo That being said, I'm also happy with not going forward with this and simply closing the PR! 😛

**Comment by jcazevedo** [comment]
> @jcazevedo That being said, I'm also happy with not going forward with this and simply closing the PR! 😛

No. Let's go with it. If it becomes more of a hassle in the future we can revisit it.

File: .github/workflows/scripts/validate_changelog.mjs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env npx zx --install
+
+import {readFileSync} from "fs";
+
+const CHANGELOG_VERSION_PREFIX = "#".repeat(2);
```
[the remark this answers is not in the document]
Comment (xRuiAlves) [reply .github/workflows/scripts/validate_changelog.mjs]: Updated in https://github.com/adzerk/apso/pull/556/commits/29f9e87e32c9208f3dd6833a905bbe9b2eac2ca9

File: .github/workflows/scripts/validate_changelog.mjs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+const CHANGELOG_VERSION_PATTERN = /^## \[\d+\.\d+\.\d+\] - \d{4}-\d{2}-\d{2}$/;
+const EXAMPLE_VERSION_LINE = "## [Unreleased]";
+const CHANGELOG_VERSION_SECTION_PREFIX = "#".repeat(3);
```
[the remark this answers is not in the document]
Comment (xRuiAlves) [reply .github/workflows/scripts/validate_changelog.mjs]: Updated in https://github.com/adzerk/apso/pull/556/commits/29f9e87e32c9208f3dd6833a905bbe9b2eac2ca9

File: .github/workflows/scripts/validate_changelog.mjs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  "Fixed",
+  "Security"
+]);
+
+const changelog = readFileSync("CHANGELOG.md").toString();
+const changelogLines = changelog.split("\n");
+
+// Validate consistency across Changelog section
+const invalidChangelogSections = changelogLines
+  .filter(line => line.startsWith(CHANGELOG_VERSION_SECTION_PREFIX))
```
[the remark this answers is not in the document]
Comment (xRuiAlves) [reply .github/workflows/scripts/validate_changelog.mjs:23]: I was assuming that **No**, we cannot

**Review by jcazevedo — APPROVED** [review]
Thanks!

**Pull request merged**